### ISSUE TITLE
Decrypting very large files via streaming

### ISSUE THREAD

**Issue by erickeniuk** [vsajip/python-gnupg#232]
**Is your feature request related to a problem? Please describe.**

Sorry, I know you get this question a lot.

I am attempting to decrypt from blob storage a 500 GiB file. I am using the  `gpg.on_data = processor` where `processor` is `processor = Streamer_5000(...)` and Streamer_5000() is below. I assumed that as I did `download_blob.readinto(stream)` that it would pull and decrypt simulataneously and thus minimize my RAM used, but it seems that I am reading the entire file into memory first, then beginning decryption. Therefore the `.on_data` does save a lot of memory but it doesn't seem to allow me to read partial files and decrypt as I go and reupload. I might just be misunderstanding streaming too.

I am picturing an even balance of bytes in -> bytes out allowing me to use a reasonable sized machine to decrypt this data (64GB RAM):

```
download_blob.readinto(stream) -> as bytes come in -> gpg.decrypt(stream,...) -> decrypted bytes -> upload_blob(chunks, ...)
```

```
class Streamer_5000:
  def __init__(self, blob_service_client, storage_account, container, blob, up_container, total_bytes):
      self.blob_service_client = blob_service_client
      self.storage_account = storage_account
      self.container = container
      self.blob =  blob
      self.up_container = up_container
      self.bytes_uploaded = 0
      self.total_bytes = total_bytes
      self.byte_block = b''

  def __call__(self, chunk):
      current_bytes = len(chunk)
      print("Byte size: ", current_bytes) # always 1024 bytes ?
      final = (current_bytes == 0)
      # Optionally overwrite container and blob from parameters
      new_name = '.'.join(self.blob.split(".")[:-1])
      # print(f"{current_bytes} bytes / chunk") reads it in in 1024 bytes (1 KiB)
      self.bytes_uploaded += current_bytes
      pf = (float(self.bytes_uploaded) / float(self.total_bytes))  * 100.0

      self.byte_block += chunk

      rss_mib = psutil.Process().memory_info().rss/(1024**2)
      mem_total_mib = psutil.virtual_memory().total/(1024**2)

      # If byte_block hits 4 MiB upload chunk to Azure blob and wipe the byte_block for next blocks
      if len(self.byte_block) == (1024**2 * 4) or final:
        print(f"! Beaming up [{current_bytes:.2f}/{total_bytes:.2f} {units}] bytes of your data stream to /{self.storage_account}/{self.container}/Decrypted/{new_name}, Scotty !\n{pf:.2f}% done\n{rss_mib:.2f}/{mem_total_mib:.2f} MiBs [{rss_mib/mem_total_mib*100:.2f}%] driver RSS Memory Used")
        container_client = self.blob_service_client.get_container_client(container=container)
        #container_client.upload_blob(name=f"Decrypted/{new_name}", data = output_data_stream, overwrite=True)
        container_client.upload_blob(name=f"Decrypted/{new_name}", data = self.byte_block, blob_type = "AppendBlob", overwrite=False)
        self.byte_block = b'' # empty byte block for next 4MiB chunk

      return False  # Tell python-gnupg not to buffer the chunk
```
I am implementing it with:
```
        input_stream = self.blob_client.download_blob(max_concurrency=12, progress_hook = self.showme)        
        processor = Streamer_5000(self.blob_service_client, self.storageAccount, container, blob, container, blob_size)
        self.gpg.on_data = processor
        time_start = time.perf_counter()
        stream = BytesIO()
        input_stream.readinto(stream)
        stream.seek(0)
        decrypted_data = self.gpg.decrypt(stream.read(), always_trust=True, passphrase=self.passphrase_secret)
```

However it seems that when I do this, it is first downloading the entire file first into memory, then it is decrypting. I have the print statement in the `progress_hook = showme` which shows how much of the file is downloaded. I also have a print statement within the `.on_data` callback. They appear as:

```
Downloading [54192.56 KiBs <-> 52.92 MiBs <-> 0.05 GiBs] size blob: [TSYZ_DATA.csv.gz.gpg] ...
Download progress: >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>-------------------- [ 32.00 / 52.92 MiBs Total Size. 60.47% done. --> DL speed: 1.0 MiB/s Estimated ETA: 0:0:52]
Download progress: >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>---------------- [ 36.00 / 52.92 MiBs Total Size. 68.02% done. --> DL speed: 1.0 MiB/s Estimated ETA: 0:0:52]
Download progress: >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>---------------- [ 36.92 / 52.92 MiBs Total Size. 69.77% done. --> DL speed: 1.0 MiB/s Estimated ETA: 0:0:52]
Download progress: >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>------------ [ 40.92 / 52.92 MiBs Total Size. 77.33% done. --> DL speed: 1.0 MiB/s Estimated ETA: 0:0:52]
Download progress: >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>-------- [ 44.92 / 52.92 MiBs Total Size. 84.88% done. --> DL speed: 1.0 MiB/s Estimated ETA: 0:0:52]
Download progress: >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>---- [ 48.92 / 52.92 MiBs Total Size. 92.44% done. --> DL speed: 1.0 MiB/s Estimated ETA: 0:0:52]
Download progress: >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> [ 52.92 / 52.92 MiBs Total Size. 100.00% done. --> DL speed: 1.0 MiB/s Estimated ETA: 0:0:52]
First 50 bytes:  b"y<1\x0e\xbdlER\xab\xb4\xe5\xea\x9a\x0fr\xf59\tw4{\xf4\xa5\xdf\x8eJ6\xfe\xe2KN\xc04\xca\x01\xfb\xdf\xbf\xbc\xe7\xed\xc2\x10'\xc8\xfat\xb6\x02\x96"
Byte size:  1024
Read chunk:  b'\x1f\x8b\x08\x08\x80\xad6d\x00\x03TSYZ_DATA.csv\x00\xcc]\xdd\x93\xda\xb8\xb2\x7f\xdf\xbfBO\xe7\xe5\x14\xac%\x7f?b\xe30d\xc0\xf8`'
Byte size:  1024

...

Read chunk:  b'1\xbf_\xb9M\x8e\xdc%\xb8\x12f\xa1\x99o\xe2x\x1b\xb6\x0b\xd2LC7\xf1\xe5\xa4\x14\xf8\x8f\xdb\xa4D\xc9\xeaF\x93\xefFi\x9a\xd6\xb8\x12\xfbt)\x04\x7fzI'
Byte size:  0
Read chunk:  b''
! Beaming up [52.92/52.92 MiBs] bytes of your data stream to /prdedssprwu2adls1/e6p/Decrypted/TSYZ_DATA.csv.gz, Scotty !
100.00% done |>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>-|
270.02/122340.00 MiBs [0.22%] driver RSS Memory Used
Decryption Successful?  True
Decryption Status:  decryption ok
```

**Describe the solution you'd like**

Really just trying to understand if I am implenting streaming via `.on_data` incorrectly or if this is correct and the entire file has to be read into memory before decrypting begins.  If true, this implies that at minimum you require the RAM the same size as the file you are decrypting with the current functionality.

If it is possible to decrypt in chunks (given I am pretty sure GPG encrypts in chunks, how could it not with RAM limitations?), perhaps an automatic feature that reads the key length and details and calculates the correct chunk size that can be decrypted incrementally or somehow stores the necessary bytes to make each chunk decrypt correctly would be interesting. Like pull the bytes out that make the cypher decrypt correctly and then strip them for each byte chunk and re-insert for each chunk where appropriate, regardless of the size (I have no clue if this would break the cipher or not, probably would).

**Describe alternatives you've considered**

I have tried `download_blob().chunk()` method too with 4 MiB but if the file is any larger than the chunk size, the decryption fails (sort of expected this since I know gpg requires some signals across the entire file to correctly decrypt, but I am not sure how to properly parse the bytes to decrypt a file successfully in chunks.

**Additional context**

**Comment by vsajip** [comment]
Decryption happens entirely within the `gpg` subprocess. The `on_data` feature allows management only of the decrypted output, which is read in chunks - and you can arrange to e,g, do custom buffering of that, or avoid buffering it altogether. Your call to `stream.read()` is what reads (downloads) the entire encrypted data into memory - to avoid this, try using `self.gpg.decrypt_file(stream, ...)` instead of `self.gpg.decrypt(stream.read(), ...)`

**Comment by erickeniuk** [comment]
@vsajip in this case if I changed it to `decrypt_file(stream)` rather than `decrypted_data = decrypt(stream.read())` would I want to remove the `gpg.on_data` and instead do the upload on the decrypted data? I tried it with `on_data` handling the data upload and it seemed to not work properly.

**Comment by vsajip** [comment]
Well, the decrypted data is fed to `on_data` if defined, otherwise buffered in memory. Using `on_data` would allow you to e.g. upload the decrypted data somewhere a chunk at a time and _not_ buffer it in memory. No idea why your `on_data` uploads didn't work properly, but it's quite unlikely to be a problem in `python_gnupg`.

**Comment by erickeniuk** [comment]
Sorry, to clarify my `.on_data` works great! Really awesome feature. The issue is that it seems that gnupg has to operate sequentially. 

1. download 100% of blob 
2. decrypt 100% of blob

If a file is say 100 GiB, it means you can potentially have ~200 GiB in memory (if I am understanding correctly). 100 GiB of the downloaded encrypted data and then the GPG decrypted data from stdout.  `.on_data` is correctly allowing me to instead pipe the `stdout` to an `upload_blob` function which is allowing me to instread do:

1. download 100% of blob
2. begin decryption
3. pipe `stdout` to `upload_blob()` in chunks (your code is 1 KiB chunks) and drop from memory the actual decrypted data as it's processed
4. finish decryption and get `.ok = True` and good `.status`

This has allowed me to take a 100 GiB file and generally stay even at ~100 GiB of RSS Mem used on the machine because I can dump the `stdout`. This is great because I am using DB cluster drivers so I can pretty easily spin up a 128 GiB RAM driver. However we have some 1 TB files and that's just not feasible or efficient.

What I am wondering is if it's possible with gnupg to read in 4 MiB / 100 GiB, decrypt the 4 MiB, and then upload that and repeat until the full 100 GiB is uploaded. I suppose this may not be possible bc it would break the cypher, but that's what I was hoping you might be able to confirm.

By the way, I was additionally wondering why you chose to `.read(1024)` in your `_read_data()` and `_copy_data()` methods? Is there something critical about doing so in 1 KiB chunks? 

I had such large files I chose to monkey patch each to read 1 GiB / chunk, given that my machine has plenty of RAM to handle a GiB of `stdout` at a time. It seems to be working fine and has reduced speeds on my outbound processing quite a bit, so I was just curious if I was missing something or if you had any additional thoughts on possibly making that variable dynamic for users with more powerful machines if it is not a critical value?

**Comment by erickeniuk** [comment]
Btw, I am happy to close this issue if there is a better avenue for this conversation. I greatly appreciate your input @vsajip

**Comment by vsajip** [comment]
> What I am wondering is if it's possible with gnupg to read in 4 MiB / 100 GiB, decrypt the 4 MiB, and then upload that and repeat until the full 100 GiB is uploaded

You're best off asking this on a GnuPG mailing list.

> By the way, I was additionally wondering why you chose to .read(1024) in your _read_data() and _copy_data() methods? Is there something critical about doing so in 1 KiB chunks?

Nothing special about 1KiB, other than it keeps `python-gnupg`'s memory usage smaller than e.g. using 1MiB buffers, I might make it configurable via a module-level attribute such as `BUFFER_SIZE`.

**Comment by vsajip** [comment]
I've now added a `buffer_size` attribute to `GPG` which will determine the buffer size for I/O between `python-gnupg` and the `gpg` executable. It should be available in version 0.5.2.

**Comment by erickeniuk** [comment]
Thank you @vsajip !
